kernel: update platform hooks documentation and usage

mathieuchopstm opened the pull request:
Update documentation related to platform hooks:
* make sure the hook functions' documentation is actually visible - currently, the dummy no-op macros are seen instead (c.f. https://docs.zephyrproject.org/latest/doxygen/html/hooks_8h.html)
* update architecture porting documentation to include a section about hooks that must be implemented in arch-specific code
  * this should reflect current practices and expectations
* clean up unnecessary `#ifdef` check in architecture files

https://builds.zephyrproject.io/zephyr/pr/97099/docs/hardware/porting/arch.html#early-boot-sequence-hooks

mathieuchopstm committed:
kernel: hooks: make definitions appear in Doxygen

The hooks' signature was not visible to Doxygen due to preprocessor
conditionals - instead, the dummy no-op implementation was seen instead.

Update #ifdef guard to make the hooks' signature and the associated
documentation visible to Doxygen such that it appears in docs.

Also fix a tiny typo on soc_reset_hook()'s documentation comment.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm committed:
doc: add hooks to arch porting documentation

Document the expected early boot sequence and when platform hooks should be
called by the arch-specific coode, and under what conditions.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm committed:
arch: *: prep_c: remove check for CONFIG_SOC_PREP_HOOK

soc_prep_hook() is always called from z_prep_c() which is implemented
as a C function. As such, there is no need to check for the associated
CONFIG_SOC_PREP_HOOK since the platform/hooks.h header will define hooks
as no-op function-like macros if their associated Kconfig isn't enabled.

Remove the Kconfig check from all arch implementations of z_prep_c() and
call soc_prep_hook() directly instead, to avoid duplicating the Kconfig
check already performed in platform/hooks.h

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

bjarki-andreasen reviewed (dismissed):
Really nice

kartben reviewed (changes requested):
very nice, thanks for this!

mathieuchopstm committed:
kernel: hooks: clean up the top-level documentation comment

Clean up the top-level comment in the platform hooks header which is used
when generating documentation:
- capitalize "soc" as "SoC"
- capitalize "zephyr"
- simplify "soc and board specific" -> "SoC/board-specific"

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm commented:
Updated per comments of @kartben (many thanks!). Also reworded the large paragraph about early reset hook to (hopefully) make it less confusing, and a few other places.

Finally, added a commit to clean up the top-level comment of the platform hooks file to get slightly better documentation :smiley: 

> reassigning, this is not a doc change anymore.

True, sorry. The `arch` editing commit was an afterthought... though it is very minor.

kartben reviewed (approved):
many thanks to _you_ @mathieuchopstm! The new documentation looks really nice and easy to read now, I believe :) https://builds.zephyrproject.io/zephyr/pr/97099/docs/hardware/porting/arch.html#early-boot-sequence-hooks
